File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/RetrieveFromMyFriendsDB.pl
#!/usr/bin/perl -w

###  RetrieveFromMyFriendsDB.pl

use strict;
use DBI;                                                              #(A)

my $dsn = "DBI:mysql:database=MyFriends;host=localhost";              #(B)
my $dbh = DBI->connect( $dsn, 'kak', 'avikak',                 
			{ RaiseError => 1, AutoCommit => 1 }  );      #(C)
print "succesfully connected \n";

my $sth = $dbh->prepare( "SHOW TABLES" );                             #(D)
$sth->execute;                                                        #(E)
my $tables_ref = $sth->fetchall_arrayref();                           #(F)
print "\nNumber of tables: " . @$tables_ref . "\n";                   #(G)
my @tabarr = @$tables_ref;                                            #(H)
print "First table: ", @{$tabarr[0]}, "\n";                           #(I)
print "Second table: ", @{$tabarr[1]}, "\n";                          #(J)
$sth->finish;                                                         #(K)

$sth = $dbh->prepare( qq{SELECT Friends.Name, Friends.Phone 
                                                FROM Friends} );      #(L)
$sth->execute;                                                        #(M)
print "\nNames and phone numbers from the Friends table:\n";
while ( my @ary = $sth->fetchrow_array ) {                            #(N)
    for ( my $i = 0; $i < @ary; $i++ ) {                              #(O)
        printf( "%s%s", $i ? "\t" : "        ", $ary[$i] );           #(P)
    }
    print "\n";
}
print "\n\n";

$sth = $dbh->prepare( q{SELECT Friends.Name, Rovers.RovingTime 
           FROM Friends, Rovers WHERE Friends.Name = Rovers.Name} );  #(Q)
$sth->execute;                                                        #(R)
print "Friends who are also in the Rovers club:\n";
my $names = $sth->{ 'NAME' };                                         #(S)
my $numFields = $sth->{ 'NUM_OF_FIELDS' };                            #(T)
for ( my $i = 0; $i < $numFields; $i++ ) {                            #(U)
    printf( "%s%s", $i ? "\t\t" : "", $$names[$i] );                  #(V)
}
print "\n";
while ( my $ref = $sth->fetchrow_arrayref ) {                         #(W)
    for ( my $i = 0; $i < $numFields; $i++ ) {                        #(X)
        printf( "%s%s", $i ? "\t" : "", $$ref[$i] );                  #(Y)
    }
    print "\n";
}
$dbh->disconnect;                                                     #(Z)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/RetrieveFromMyFriendsDB.py
#!/usr/bin/env python

###  RetrieveFromMyFriendsDB.py

import sys                                                           #(A)
import MySQLdb                                                       #(B)

try:                                                                 #(C)
    conn = MySQLdb.connect( host='localhost',                   
                            user='kak',
                            passwd='avikak',
                            db='MyFriends')                          #(D)
except MySQLdb.Error, e:                                             #(E)
    print "Unable to connect to database. Error %d: %s" % \
                                  (e.args[0], e.args[1])             #(F)
    sys.exit(1)                                                      #(G)
    
try:                                                                 #(I)
    cursor = conn.cursor()                                           #(J)

    cursor.execute( "SHOW TABLES" )      # Friends                   #(K)
                                         # Rovers
    result = cursor.fetchall()                                       #(L)
    for record in result:                                            #(M)
        print "%s" % record[0]                                       #(N)

    cursor.execute("SELECT Friends.Name, Friends.Phone FROM Friends")#(O)
    result = cursor.fetchall()                                       #(P)
    for record in result:                                            #(Q)
        print "%s\t%s" % (record[0], record[1])                      #(R)
                                         # Ziggy Zaphod    4569876
                                         # Yo Yo Ma        3472828
                                         # Gogo Gaga       27278927

    cursor.execute("SELECT Friends.Name, Rovers.RovingTime \
           FROM Friends, Rovers WHERE Friends.Name = Rovers.Name" )  #(S)
    result = cursor.fetchall()                                       #(T)
    for record in result:                                            #(U)
        print "%s\t%s" % (record[0], record[1])                      #(V)
                                         # Yo Yo Ma      8 pm
    cursor.close()                                                   #(W)

except MySQLdb.ProgrammingError, e:                                  #(X)
    print "Possible SQL Syntax Error %d: %s" % (e.args[0], e.args[1])#(Y)
except MySQLdb.Error, e:                                             #(Z)
    print "Error %d: %s" % (e.args[0], e.args[1])                    #(a)
    sys.exit(1)                                                      #(b)
    
conn.close()                                                         #(c)
